great_expectations/experimental/metric_repository/metric_list_retriever.py

```diff
@@ -0,0 +1,322 @@
+logging.captureWarnings(True)
+
+_MetricKey: TypeAlias = Union[Tuple[str, Hashable, Hashable], Tuple[str, str, str]]
+
+
+class MetricListMetricRetriever(MetricRetriever):
+    def __init__(self, context: AbstractDataContext):
+        super().__init__(context=context)
+        self._validator: Validator | None = None
+
+    @override
+    def get_metrics(
+        self,
+        batch_request: BatchRequest,
+        metric_list: Optional[List[MetricTypes]] = None,
+    ) -> Sequence[Metric]:
+        metrics_result: List[Metric] = []
+
+        if not metric_list:
+            raise ValueError("metric_list cannot be empty")
+
+        self._check_valid_metric_types(metric_list)
+
+        table_metrics = self._get_table_metrics(
+            batch_request=batch_request, metric_list=metric_list
+        )
+        metrics_result.extend(table_metrics)
+
+        # exit early if only Table Metrics exist
+        if not self._column_metrics_in_metric_list(metric_list):
+            return metrics_result
+
+        table_column_types = list(
+            filter(
+                lambda m: m.metric_name == MetricTypes.TABLE_COLUMN_TYPES, table_metrics
+            )
+        )[0]
+
+        # We need to skip columns that do not report a type, because the metric computation
+        # to determine semantic type will fail.
+        exclude_column_names = self._get_columns_to_exclude(table_column_types)
+
+        numeric_column_names = self._get_numeric_column_names(
+            batch_request=batch_request, exclude_column_names=exclude_column_names
+        )
+        timestamp_column_names = self._get_timestamp_column_names(
+            batch_request=batch_request, exclude_column_names=exclude_column_names
+        )
+        numeric_column_metrics = self._get_numeric_column_metrics(
+            metric_list, batch_request, numeric_column_names
+        )
+        timestamp_column_metrics = self._get_timestamp_column_metrics(
+            metric_list, batch_request, timestamp_column_names
+        )
+        all_column_names: List[str] = self._get_all_column_names(table_metrics)
+        non_numeric_column_metrics = self._get_non_numeric_column_metrics(
+            metric_list, batch_request, all_column_names
+        )
+
+        bundled_list = list(
+            chain(
+                table_metrics,
+                numeric_column_metrics,
+                timestamp_column_metrics,
+                non_numeric_column_metrics,
+            )
+        )
+
+        return bundled_list
+
+    def _get_non_numeric_column_metrics(
+        self,
+        metrics_list: List[MetricTypes],
+        batch_request: BatchRequest,
+        column_list: List[str],
+    ) -> Sequence[Metric]:
+        column_metric_names = {MetricTypes.COLUMN_NULL_COUNT}
+        metrics: list[Metric] = []
+        metrics_list_as_set = set(metrics_list)
+        metrics_to_calculate = sorted(
+            column_metric_names.intersection(metrics_list_as_set)
+        )
```

Comment:
How should this work if the metric isn't supported for the column type? E.g. if `metrics_list` contains a min or max metric for a string column?

Reply:
update: talked this out, and I think we're doing the right thing here. Will add a docstring